zcash_client_sqlite/src/testing.rs

```diff
@@ -1938,22 +1937,62 @@ impl BlockCache {
     }
 }
 
+pub(crate) struct NoteCommitments {
+    sapling_note_commitments: Vec<sapling::Node>,
+    #[cfg(feature = "orchard")]
+    orchard_note_commitments: Vec<MerkleHashOrchard>,
+}
+
+impl NoteCommitments {
+    #[allow(dead_code)]
+    pub(crate) fn sapling_note_commitments(&self) -> &[sapling::Node] {
+        self.sapling_note_commitments.as_ref()
+    }
+
+    #[cfg(feature = "orchard")]
+    pub(crate) fn orchard_note_commitments(&self) -> &[MerkleHashOrchard] {
+        self.orchard_note_commitments.as_ref()
+    }
+}
+
 impl TestCache for BlockCache {
     type BlockSource = BlockDb;
-    type InsertResult = ();
+    type InsertResult = NoteCommitments;
 
     fn block_source(&self) -> &Self::BlockSource {
         &self.db_cache
     }
 
-    fn insert(&self, cb: &CompactBlock) {
+    fn insert(&self, cb: &CompactBlock) -> Self::InsertResult {
         let cb_bytes = cb.encode_to_vec();
+        let res = NoteCommitments {
+            sapling_note_commitments: cb
+                .vtx
+                .iter()
+                .flat_map(|tx| {
+                    tx.outputs
+                        .iter()
+                        .map(|out| sapling::Node::from_cmu(&out.cmu().unwrap()))
+                })
+                .collect(),
+            #[cfg(feature = "orchard")]
+            orchard_note_commitments: cb
+                .vtx
+                .iter()
+                .flat_map(|tx| {
+                    tx.actions
+                        .iter()
+                        .map(|act| MerkleHashOrchard::from_cmx(&act.cmx().unwrap()))
+                })
+                .collect(),
+        };
```

Comment:
Move this into a private constructor `NoteCommitments::from_compact_block` (enabling its reuse as necessary), or move its parts onto `CompactBlock` itself and remove `NoteCommitments`.